packages/cli/lib/__tests__/projects.test.js

```diff
@@ -0,0 +1,153 @@
+const fs = require('fs');
+const os = require('os');
+const path = require('path');
+const { EXIT_CODES } = require('../enums/exitCodes');
+const projects = require('../projects');
+
+describe('@hubspot/cli/lib/projects', () => {
+  describe('validateProjectConfig()', () => {
+    let realProcess;
+    let projectDir;
+    let exitMock;
+    let errorSpy;
+
+    beforeAll(() => {
+      projectDir = fs.mkdtempSync(path.join(os.tmpdir(), 'projects-'));
+      fs.mkdirSync(path.join(projectDir, 'src'));
```

Comment:
TIL about `mkdtempSync`. Do these temp folders remove themselves if the tests fail to complete for some reason? There are scenarios where the after lifecycle methods never run, which could cause the folders to be left around in the project. Can you do this without actually creating the temp folder? It looks like the only fs method that might require the folder to actually exist is  [existsSync](https://github.com/HubSpot/hubspot-cli/pull/951/files#diff-fa4fa15b34bd2259fc2c24ec990fdbaa6492e6030d940700895318b399058174R188), so you could mock that instead and avoid having to write anything to the filesystem.

Thanks for adding tests btw! 😄

Reply:
Yeah avoiding mocking wherever possible is a good rule of thumb. For filesystem methods though we typically try to avoid ever actually writing because it can potentially have an impact on the repo structure depending on what the utils are doing. I did a little reading and it looks like [the better solution is to use memfs ](https://medium.com/nerd-for-tech/testing-in-node-js-easy-way-to-mock-filesystem-883b9f822ea4)(not currently a dep of the CLI though and probably out of the scope of this PR).

The main issue I'm trying to avoid is someone running these tests locally, unintentionally modifying the project structure, and then committing test files ([I accidentally did this last week](https://github.com/HubSpot/hubspot-local-dev-lib/pull/54#pullrequestreview-1708649549)). Now that I've read that article I'm realizing we should switch up the approach at some point, but regardless the current pattern we've been using is to mock fs to prevent any actual writes.